src/task-dispatcher/td-lean.hh

```diff
@@ -234,6 +235,27 @@ void submit_lambda_on_counter(handle::counter handle, F&& func)
 
     submit_on_counter(handle, &dispatch, 1);
 }
+
+template <class F>
+uint32_t submit_batched_on_counter(handle::counter handle, F&& func, uint32_t num_elements, uint32_t max_num_batches, cc::allocator* scratch = cc::system_allocator)
+{
+    static_assert(std::is_invocable_v<F, uint32_t, uint32_t, uint32_t>, "function must be invocable with element start, end, and index argument");
+
+    auto const batch_size = cc::int_div_ceil(num_elements, max_num_batches);
+    auto const num_batches = cc::int_div_ceil(num_elements, batch_size);
+    CC_ASSERT(num_batches <= max_num_batches && "programmer error");
+
+    auto tasks = cc::alloc_array<td::container::task>::uninitialized(num_batches, scratch);
+
+    for (uint32_t batch = 0u, start = 0u, end = cc::min(batch_size, num_elements); //
+         batch < num_batches;                                                      //
+         ++batch, start = batch * batch_size, end = cc::min((batch + 1) * batch_size, num_elements))
+    {
+        tasks[batch].lambda([=] { func(start, end, batch); });
```

Comment:
is this by design that the func is copied?

Reply:
Yes, the task is persisted, the original lambda will be dead once it's executed